commands/operator-sdk/cmd/scorecard/test_definitions.go

```diff
@@ -0,0 +1,277 @@
+	TestInfo
+	OLMTestConfig
+}
+
+func NewAnnotationsContainExamplesTest(conf OLMTestConfig) *AnnotationsContainExamplesTest {
+	return &AnnotationsContainExamplesTest{
+		OLMTestConfig: conf,
+		TestInfo: TestInfo{
+			Name:        "CRs have at least 1 example",
+			Description: "The CSV's metadata contains an alm-examples section",
+		},
+	}
+}
+
+type SpecDescriptorsTest struct {
+	TestInfo
+	OLMTestConfig
+}
+
+func NewSpecDescriptorsTest(conf OLMTestConfig) *SpecDescriptorsTest {
+	return &SpecDescriptorsTest{
+		OLMTestConfig: conf,
+		TestInfo: TestInfo{
+			Name:        "Spec fields with descriptors",
+			Description: "All spec fields have matching descriptors in the CSV",
+		},
+	}
+}
+
+type StatusDescriptorsTest struct {
+	TestInfo
+	OLMTestConfig
+}
+
+func NewStatusDescriptorsTest(conf OLMTestConfig) *StatusDescriptorsTest {
+	return &StatusDescriptorsTest{
+		OLMTestConfig: conf,
+		TestInfo: TestInfo{
+			Name:        "Status fields with descriptors",
+			Description: "All status fields have matching descriptors in the CSV",
+		},
+	}
+}
+
+// Test Suite Declarations
+
+func NewBasicTestSuite(conf BasicTestConfig) *TestSuite {
+	ts := NewTestSuite(
+		"Basic Tests",
+		"Test suite that runs basic, functional operator tests",
+	)
+	ts.AddTest(NewCheckSpecTest(conf), 44)
+	ts.AddTest(NewCheckStatusTest(conf), 28)
+	ts.AddTest(NewWritingIntoCRsHasEffectTest(conf), 28)
+
+	return ts
+}
+
+func NewOLMTestSuite(conf OLMTestConfig) *TestSuite {
+	ts := NewTestSuite(
+		"OLM Tests",
+		"Test suite checks if an operator's CSV follows best practices",
+	)
+
+	ts.AddTest(NewCRDsHaveValidationTest(conf), 24)
+	ts.AddTest(NewCRDsHaveResourcesTest(conf), 19)
+	ts.AddTest(NewAnnotationsContainExamplesTest(conf), 19)
+	ts.AddTest(NewSpecDescriptorsTest(conf), 19)
+	ts.AddTest(NewStatusDescriptorsTest(conf), 19)
+
+	return ts
+}
+
+// Helper functions
+
+// ResultsPassFail will be used when multiple CRs are supported
+func ResultsPassFail(results []TestResult) (earned, max int) {
+	for _, result := range results {
+		if result.EarnedPoints != result.MaximumPoints {
+			return 0, 1
+		}
+	}
+	return 1, 1
+}
+
+// ResultsCumulative will be used when multiple CRs are supported
+func ResultsCumulative(results []TestResult) (earned, max int) {
+	for _, result := range results {
+		earned += result.EarnedPoints
+		max += result.MaximumPoints
+	}
+	return earned, max
+}
+
+func (ts *TestSuite) AddTest(t Test, weight int) {
+	ts.Tests = append(ts.Tests, t)
+	ts.Weights[t.GetName()] = weight
+}
+
+func (ts *TestSuite) TotalScore() (score int) {
+	floatScore := 0.0
+	for _, result := range ts.TestResults {
+		if result.MaximumPoints != 0 {
+			floatScore += (float64(result.EarnedPoints) / float64(result.MaximumPoints)) * float64(ts.Weights[result.Test.GetName()])
+		}
+	}
+	return int(floatScore)
+}
+
+func (ts *TestSuite) Run(ctx context.Context) {
+	for _, test := range ts.Tests {
+		ts.TestResults = append(ts.TestResults, test.Run(ctx))
+	}
+}
```

Comment:
How are we thinking the multiple CR scorecard testing will work? Are we thinking we'll loop over each CR and run all of the enabled tests for that CR during each iteration? Or are we thinking we'll setup the tests to be multi-CR aware?

If the latter, it might be worth making `TestSuite` implement the `Test` interface (by having the correct `Run()` signature) and looking into whether we can use `TestSuite` to build a suite that executes the same test against a slice of CRs. That's kind of what I was getting at in my [previous comment](https://github.com/operator-framework/operator-sdk/pull/994#discussion_r253901901) about suites of suites.

Reply:
The way I was thinking of implementing it was looping through the tests once per CR and storing all the results. To get total scores for each test, we can then use the `ResultsCumulative` and `ResultsPassFail` functions, depending on whether or not the `isCumulative` bool is set. I'll make a helper function in a future PR to handle combining the results of multiple runs of a suite.

Reply:
> To get total scores for each test, we can then use the `ResultsCumulative` and `ResultsPassFail` functions, depending on whether or not the `isCumulative` bool is set.

Maybe I'm getting ahead of myself, since the multi-CR support is not in scope for this PR, but with `isCumulative` being a field on each TestResult, it's not apparent to me how we would decide which `TestResult`s to group together in a call to `ResultsCumulative` or `ResultsPassFail`.

I'm probably missing something, but it seems like `isCumulative` would be associated with a set of tests and results, not an individual test result, since there's no context at that level about what other results should be included. That's why I was thinking that the TestSuite would have a scoring function that applied to its group of tests. Am I thinking about this incorrectly?

Reply:
It would probably make more sense to put `isCumulative` into `TestInfo`. Certain tests (like `specDescriptors`) are meant to have the scores of each run added together, resulting in a score such as `3/5`. Other tests (like `checkSpec`) are meant to either pass or fail. For instance, if 2 out of 3 CRs have spec blocks, the overall score for `checkSpec` is a `0/1`. At the moment, it seems that all the pass/fail tests are in the Basic suite and all the cumulative tests are in the OLM suite, but that might change as more tests are added.

So it's not the suites that have cumulative or non-cumulative scores, it's the tests themselves that are.

Reply:
Ahh okay. Makes sense. Thanks for the explanation :+1: